File: android/src/main/java/org/maplibre/reactnative/components/styles/MLRNStyleFunctionParser.java
package org.maplibre.reactnative.components.styles;

import android.util.Log;

import com.facebook.react.bridge.Dynamic;
import com.facebook.react.bridge.NoSuchKeyException;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableNativeArray;
import com.facebook.react.bridge.ReadableType;
import org.maplibre.android.style.expressions.Expression.Stop;
import org.maplibre.android.style.layers.PropertyValue;

import java.text.NumberFormat;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class MLRNStyleFunctionParser<T, V> {
    private MLRNStyleValue mStyleValue;

    public MLRNStyleFunctionParser(MLRNStyleValue styleValue) {
        mStyleValue = styleValue;
    }

    public List<StopConfig> getRawStops() {
        ReadableArray readableArrayRawStops = mStyleValue.getArray("stops");

        List<StopConfig> rawStops = new ArrayList<>();

        for (int i = 0; i < readableArrayRawStops.size(); i++) {
            ReadableArray rawStop = readableArrayRawStops.getArray(i);

            ReadableMap jsStopKey = rawStop.getMap(0);
            ReadableMap jsStopValue = rawStop.getMap(1);
            MLRNStyleValue innerStyleValue = new MLRNStyleValue(jsStopValue);

            Object propertyValue = null;
            try {
                Dynamic dynamicPropertyValue = innerStyleValue.getDynamic("propertyValue");
                ReadableType type = dynamicPropertyValue.getType();

                if (type.equals(ReadableType.Number)) {
                    propertyValue = dynamicPropertyValue.asDouble();
                } else if (type.equals(ReadableType.Boolean)){
                    propertyValue = dynamicPropertyValue.asBoolean();
                } else {
                    propertyValue = dynamicPropertyValue.asString();
                }
            } catch (NoSuchKeyException e) {
                // not a zoom-property value
            }

            StopConfig config;
            if (propertyValue != null) {
                config = new StopConfig(getStopKey(jsStopKey), getRawStopValue(innerStyleValue), propertyValue);
            } else {
                config = new StopConfig(getStopKey(jsStopKey), getRawStopValue(innerStyleValue));
            }

            rawStops.add(config);
        }

        return rawStops;
    }

    protected abstract T getRawStopValue (MLRNStyleValue styleValue);
    protected abstract PropertyValue<V> getStopValue(T value);

    private Object getStopKey(ReadableMap jsStopKey) {
        String payloadKey = "value";
        String type = jsStopKey.getString("type");

        switch (type) {
            case "number":
                return jsStopKey.getDouble(payloadKey);
            case "boolean":
                return jsStopKey.getBoolean(payloadKey);
            default:
                return jsStopKey.getString(payloadKey);
        }
    }

    private class StopConfig {
        Object propertyValue;
        T value;
        Object key;

        StopConfig(Object key, T value) {
            this(key, value, null);
        }

        StopConfig(Object key, T value, Object propertyValue) {
            this.key = key;
            this.value = value;
            this.propertyValue = propertyValue;
        }
    }
}

File: package/android/src/main/java/org/maplibre/reactnative/components/layer/style/MLRNStyleFunctionParser.kt
package org.maplibre.reactnative.components.layer.style

import com.facebook.react.bridge.NoSuchKeyException
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import org.maplibre.android.style.layers.PropertyValue

abstract class MLRNStyleFunctionParser<T, V>(
    private val mStyleValue: MLRNStyleValue,
) {
    val rawStops: MutableList<MLRNStyleFunctionParser<T, V>.StopConfig?>
        get() {
            val readableArrayRawStops = mStyleValue.getArray("stops")

            val rawStops: MutableList<StopConfig?> =
                ArrayList<StopConfig?>()

            if (readableArrayRawStops != null) {
                for (i in 0..<readableArrayRawStops.size()) {
                    val rawStop = readableArrayRawStops.getArray(i)

                    val jsStopKey = rawStop!!.getMap(0)
                    val jsStopValue = rawStop.getMap(1)
                    val innerStyleValue = MLRNStyleValue(jsStopValue!!)

                    var propertyValue: Any? = null
                    try {
                        val dynamicPropertyValue =
                            innerStyleValue.getDynamic("propertyValue")
                        val type = dynamicPropertyValue.type

                        if (type == ReadableType.Number) {
                            propertyValue = dynamicPropertyValue.asDouble()
                        } else if (type == ReadableType.Boolean) {
                            propertyValue = dynamicPropertyValue.asBoolean()
                        } else {
                            propertyValue = dynamicPropertyValue.asString()
                        }
                    } catch (e: NoSuchKeyException) {
                        // not a zoom-property value
                    }

                    val config: StopConfig?
                    if (propertyValue != null) {
                        config =
                            StopConfig(
                                getStopKey(jsStopKey!!),
                                getRawStopValue(innerStyleValue),
                                propertyValue,
                            )
                    } else {
                        config =
                            StopConfig(getStopKey(jsStopKey!!), getRawStopValue(innerStyleValue))
                    }

                    rawStops.add(config)
                }
            }

            return rawStops
        }

    protected abstract fun getRawStopValue(styleValue: MLRNStyleValue?): T?

    protected abstract fun getStopValue(value: T?): PropertyValue<V?>?

    private fun getStopKey(jsStopKey: ReadableMap): Any? {
        val payloadKey = "value"
        val type = jsStopKey.getString("type")

        when (type) {
            "number" -> return jsStopKey.getDouble(payloadKey)
            "boolean" -> return jsStopKey.getBoolean(payloadKey)
            else -> return jsStopKey.getString(payloadKey)
        }
    }

    inner class StopConfig
        constructor(
            var key: Any?,
            var value: T?,
            var propertyValue: Any? = null,
        )
}
